### DISCUSSION TITLE
Fixes for 8.0.0

### DISCUSSION THREAD

**Discussion by philipphofmann** [getsentry/sentry-cocoa#1094] [General]
Ideas for fixes in 8.0.0.

**Comment by philipphofmann** [comment]
Rename `SentryDebugMeta` to `SentryDebugImage` and add a new `SentryDebugMeta` that aligns with the [DebugMeta](https://develop.sentry.dev/sdk/event-payloads/debugmeta) from the unified API.

**Comment by armcknight** [comment]
Mark some properties in SentryEvent.h `readonly` to avoid pointer modification by consumers. Then we can remove the scary warnings in the headerdocs like this: https://github.com/getsentry/sentry-cocoa/blob/ce1b8d1375ddfc044fdc60e9a5c3505f9b80ab86/Sources/Sentry/Public/SentryEvent.h#L62-L63

**Comment by armcknight** [comment]
`SentryTransactionContext` is the only subclass of `SentrySpanContext`. They should be combined into a single class.

**Reply by armcknight** [reply]
Same with `SentryTransaction`, which is the sole subclass of `SentryEvent`.

**Reply by armcknight** [reply]
Given both of these I'd say we should end with a `SentryTransaction` (subsuming `SentryEvent`) and `SentryTransactionContext` (subsuming `SentrySpanContext`).

**Reply by kevinrenskers** [reply]
`SentryTransactionContext` is the only subclass of `SentrySpanContext`, but they are both being used, right? So combining them into a single class would add properties to `SentrySpanContext` that maybe shouldn't be there?

**Reply by armcknight** [reply]
Yes, you are right, SentrySpanContext is used in `-[SentryTracer startChildWithParentId...]`. I can't tell from the various places they're both used how they overlap or differ, so this can just stay as is.

**Comment by armcknight** [comment]
`SentrySamplingContext.customSamplingContext` is never read by anything to send to the backend, nor to alter the SDK's behavior, so should be removed.

**Reply by kevinrenskers** [reply]
It can be used by the user of the SDK by setting `profilesSampler` in SentryOptions. This is used to dynamically determine the sample rate, and that function can then use the custom context. But wow that is a lot of code involved just for this one feature 😅

**Reply by armcknight** [reply]
I see it now, thanks for pointing that out. It looks like `customSamplingContext` always appears alongside a `SentryTransactionContext`, which I assume is also the same one a `SentrySamplingContext` references, and consumers that implement `tracesSampler` have access to the transaction context via the property on the sampling context, so they probably _could_ be combined, but it's not a huge deal.

**Comment by armcknight** [comment]
We should not expose SentryOptions.diagnosticLevel for public consumption/modification. Logging should be either info/error as the default, or debug of the SDK debug mode is enabled.

**Reply by kevinrenskers** [reply]
Strongly disagree with this. "Logging should be either info/error as the default" - therein lies the problem. Which one do you pick? Neither one will be the right choice for 100% of people. And even while working on an app in debug mode, the debug level can simply be too noisy. I really think this needs to be configurable by users of the SDK. It's a fundamental option that basically all SDKs offer.

**Comment by armcknight** [comment]
We code `SentryBreadcrumb.type` as an `NSString` with a comment that it should only be one of a given set of values

> Type of breadcrumb, can be e.g.: http, empty, user, navigation

However we have other types listed at https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/#breadcrumb-types

Instead of leaving this seemingly open-ended (what happens if consumers provide a value that's not in the list? The above doc doesn't say, nor does https://docs.sentry.io/platforms/apple/enriching-events/breadcrumbs/) with a comment on what is valid (can't be automatically checked and drifts from implementation), we should provide an enum of valid values.

We should probably also validate that the data being provided to a breadcrumb matches the type of breadcrumb it is, logging a warning for any unexpected field keys.

**Reply by armcknight** [reply]
Just checked [the Android SDK](https://github.com/getsentry/sentry-java/blob/7dd32c022ab7acbf24fba1253b4ecb11a5d8c9af/sentry/src/main/java/io/sentry/Breadcrumb.java#L25) to see if it's done differently in another SDK and it is implemented the same way, so I'm digging for further requirements on this. See https://github.com/getsentry/sentry/issues/37674

**Comment by kevinrenskers** [comment]
When are we planning to start on version 8 anyway? We have a few issues tagged with 8.0.0, we have identified some good improvements here, we'd love to bump the minimum OS versions, etc. Version 7.0.0 was released in April 2021, so maybe we can start to think about starting on 8 soon?

**Reply by philipphofmann** [reply]
Yes, there are a few issues tagged with 8.0.0, but is there a significant one worth shipping right now? I rather wait as long as we can to avoid breaking changes.

**Comment by brustolin** [comment]
Include Swift support in the SDK.

**Reply by kevinrenskers** [reply]
What do you mean with this? We already support apps built with Swift, so I am not sure what "Swift support" means? In https://github.com/getsentry/sentry-cocoa/issues/2079 you mentioned SwiftUI which is a good example, but would be good to document it here.

**Reply by brustolin** [reply]
I tried this, and it does not work without project refactoring. Not sure if this worth.

**Reply by armcknight** [reply]
Would this be a breaking change to any current API @brustolin ?

**Reply by philipphofmann** [reply]
That will be done with https://github.com/getsentry/sentry-cocoa/pull/2162

**Comment by armcknight** [comment]
**Refactor our usages of `NSDate`**. This has come up when making some changes for profiling, which uses `clock_gettime_nsec_np` and falls back to `mach_absolute_time` for earlier OS versions. Those aren't (easily) compatible with `NSDate`.

We do have `SentryCurrentDate`, `SentryCurrentDateProvider` and `SentryDefaultCurrentDateProvider` that can return a `dispatch_time_t`, but almost all of our timekeeping uses NSDate, for span/event/etc start/stop and computing durations. However, `NSDate` is not a great API for this since it can change at any time, including for time server syncs or when end users change time settings on their devices.

For my current changes I'm just adding a system timestamp alongside the NSDates where needed, but we could refactor this in the future into something like this:

```objective-c
@interface SentryDate : NSObject

/** The @c NSDate reported when this instance was initialized. */
@property (strong, nonatomic) NSDate *nsdate;

/** The system clock time reported when this instance was initialized. */
@property (assign, nonatomic) uint64_t systemTime;
@end
```

We could also provide methods to calculate durations and compare two dates, and we can check if the calculation differs too greatly between using the NSDate and system time.

I started down this road but wound up having to modify `SentryAppStartMeasurement` which is a public header.

**Reply by armcknight** [reply]
So far this comes into play when using `startTimestamp`/`timestamp` on spans/events/etc, as well as `CFTimeInterval`s used in `SentryFramesTracker`.

**Comment by armcknight** [comment]
Remove deprecated arrays of enum name constants in favor of functions to map enum values to names (see [this PR](https://github.com/getsentry/sentry-cocoa/pull/2108)).

**Comment by philipphofmann** [comment]
Mark close as deprecated, use close(timeout:) instead, which calls flush. This change aligns with the unified API https://develop.sentry.dev/sdk/unified-api/#client.

**Comment by armcknight** [comment]
**Bump minimum supported versions of macOS/iOS/etc**. Note that we currently don't test all the oldest versions we support, per https://github.com/getsentry/sentry-cocoa/blob/4a66f00b1d9816d03fc915336a36afcc21e2d526/.github/workflows/test.yml#L62-L67
See also https://github.com/getsentry/sentry-cocoa/issues/2218.

GitHub Actions has deprecated the macOS 10.15 image providing Xcode 10 and will no longer support it starting 1 Dec 2022 ([source](https://github.com/actions/runner-images/issues/5583)).

We currently support down to macOS 10.10, iOS/tvOS 9.0 and watchOS 2.0.

The next lowest image supported by GHA after the 10.15 EOL is macOS 11 which would be capable of running down to Xcode 11.7 and thus supporting down to iOS/tvOS 13.7 and watchOS 6.2.

We could rely on SauceLabs (which offers an iPhone down to iOS 10.3.2) or similar to bridge the gap, but we have not had success running older devices.

**Reply by philipphofmann** [reply]
Will be done with https://github.com/getsentry/sentry-cocoa/issues/1916

**Comment by philipphofmann** [comment]
Remove [`SentrySDK.startWithOptions:dict`](https://github.com/getsentry/sentry-cocoa/blob/791123de3f8fc898cc93f3bd5ffb8bbb213681e5/Sources/Sentry/Public/SentrySDK.h#L33). Only [RN](https://github.com/getsentry/sentry-react-native/blob/7122f76aac3d6887418f9c57489f5fb446bd8f15/ios/RNSentry.m#L67-L102) uses it, and they could use the same approach as on [Flutter](https://github.com/getsentry/sentry-dart/blob/fa7e619df10918bff2062545c9746df886c878bf/flutter/ios/Classes/SentryFlutterPluginApple.swift#L251).

We have to write a lot of code to keep supporting the [`SentrySDK.startWithOptions:dict`] https://github.com/getsentry/sentry-cocoa/blob/791123de3f8fc898cc93f3bd5ffb8bbb213681e5/Sources/Sentry/SentryOptions.m#L169-L327

**Reply by philipphofmann** [reply]
@kevinrenskers will take care https://github.com/getsentry/sentry-cocoa/issues/2395.

**Comment by armcknight** [comment]
Closing this as major version 8 has been released. See the new discussion for ideas for next major version 9: https://github.com/getsentry/sentry-cocoa/discussions/2946
